internal/interpreter/highlight.rs

```diff
@@ -3,22 +3,171 @@
+
+        if next == *item {
+            return next;
+        }
+        let offset = next.map_to_window(LogicalPoint::default());
+        let geometry = next.geometry().translate(offset.to_vector());
+
+        if geometry.contains(*position) {
+            return next;
+        }
+    }
+}
+
+fn element_providing_item(component: &DynamicComponentVRc, index: usize) -> Option<ElementRc> {
+    generativity::make_guard!(guard);
+    let c = component.unerase(guard);
+
+    return c.description().original_elements.get(index).cloned();
+}
+
+fn map_offset_to_line(
+    source_file: Option<SourceFile>,
+    start_offset: usize,
+    end_offset: usize,
+) -> (String, u32, u32, u32, u32) {
+    if let Some(sf) = source_file {
+        let file_name = sf.path().to_string_lossy().to_string();
+        let (start_line, start_column) = sf.line_column(start_offset);
+        let (end_line, end_column) = sf.line_column(end_offset);
+        return (
+            file_name,
+            start_line as u32,
+            start_column as u32,
+            end_line as u32,
+            end_column as u32,
+        );
+    } else {
+        return (String::new(), 0, 0, 0, 0);
+    }
+}
+
+struct DesignModeState {
+    pub current_item: Option<ItemWeak>,
+}
+
+pub fn set_current_element_information_callback(
+    component_instance: &DynamicComponentVRc,
+    callback: Box<dyn Fn(String, u32, u32, u32, u32) -> ()>,
+) {
+    let weak_component = VRc::downgrade(component_instance);
+
+    generativity::make_guard!(guard);
+    let c = component_instance.unerase(guard);
+
+    let _ = c.description().set_callback_handler(
+        c.borrow(),
+        CURRENT_ELEMENT_CALLBACK_PROP,
+        Box::new(move |values: &[Value]| -> Value {
+            static mut STATE: DesignModeState = DesignModeState { current_item: None };
+
+            let position = LogicalPoint::new(
+                if let Some(Value::Number(n)) = values.get(0) { *n as f32 } else { f32::MAX },
+                if let Some(Value::Number(n)) = values.get(1) { *n as f32 } else { f32::MAX },
+            );
+
+            let c = if let Some(c) = weak_component.upgrade() {
+                c
+            } else {
+                callback(String::new(), 0, 0, 0, 0);
+                return Value::Void;
+            };
+
+            let start_item = unsafe { STATE.current_item.take() }
+                .and_then(|i| i.upgrade())
+                .unwrap_or_else(|| ItemRc::new(VRc::into_dyn(c.clone()), 0));
+
+            let i = find_item(&start_item, &position);
+            unsafe { STATE.current_item = Some(i.downgrade()) };
+
+            if let Some((file, start_line, start_column, end_line, end_column)) =
+                element_providing_item(
+                    unsafe {
+                        std::mem::transmute::<
+                            &VRc<ComponentVTable>,
+                            &VRc<ComponentVTable, ErasedComponentBox>,
+                        >(&i.component())
```

Comment:
There must be a safe way to do that.
I think you can use VRc::downgrade

This is otherwise not safe at all since maybe the VRc comes from some other stuff than ErasedComponentBox once we start mixing native and dynamic UI.

Reply:
Using `VRc::downgrade` should work.

Reply:
Ah it is VRef::downcast